fetcher/src/job/job.utils.ts

```diff
@@ -92,23 +117,55 @@ function validateAdapter(adapter): IAdapter {
   }
 }
 
+function selectProxy(url: string, proxies) {
+  const source = url.split('/')[2]
+  const proxySize = proxies.length
+
+  if (!source) {
+    throw new FetcherError(FetcherErrorCode.InvalidUrl)
+  }
+  if (proxySize == 0) {
+    throw new FetcherError(FetcherErrorCode.UnexpectedNumberOfProxies)
+  }
+
+  if (source in proxySourceMap) {
+    proxySourceMap[source] = (proxySourceMap[source] + 1) % proxySize
+  } else {
+    proxySourceMap[source] = latestId
+    latestId = (latestId + 1) % proxySize
+  }
+
+  return proxies[proxySourceMap[source]]
+}
+
 export function extractFeeds(
   adapter,
   aggregatorId: bigint,
   aggregatorHash: string,
   threshold: number,
   absoluteThreshold: number,
-  address: string
+  address: string,
+  proxies,
+  logger
 ) {
   const adapterHash = adapter.adapterHash
   const feeds = adapter.feeds.map((f) => {
+    let proxy = {}
```

Comment:
I think the `{}` represents no proxy, however, it would fail in subsequent calls where you directly access `host` and `port` properties without any check whether there is anything.

It would be better to define type for proxy

```typescript
type Proxy = {
  host: string | undefined;
  port: number | undefined;
};
```

and then just define it with type

```typescript
let proxy: Proxy
```

At places where you use it, you should check whether it is `undefined` before use.